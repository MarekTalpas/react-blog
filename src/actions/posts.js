import database from '../firebase/firebase';

const addPost = post => ({
  type: 'ADD_POST',
  post
});

const initiateAddPost = (postData = {}) => {
  return (dispatch, getState) => {
    const {
      title = '',
      content = '',
      createdAt = 0
    } = postData;
    const post = { title, content, createdAt };
    const { uid } = getState().auth;

    return database.ref(`users/${uid}/posts`).push(post).then(({ key: id }) => {
      dispatch(addPost({
        id,
        ...post
      }));
    });
  };
};

const removePost = ({ id } = {}) => ({
  type: 'REMOVE_POST',
  id
});

const initiateRemovePost = ({ id } = {}) => {
  return (dispatch, getState) => {
    const { uid } = getState().auth;

    return database.ref(`users/${uid}/posts/${id}`).remove().then(() => {
      dispatch(removePost({ id }));
    });
  };
};

const editPost = (id, update) => ({
  type: 'EDIT_POST',
  id,
  update
});

const initiateEditPost = (id, update) => {
  return (dispatch, getState) => {
    const { uid } = getState().auth;

    return database.ref(`users/${uid}/posts/${id}`).update(update).then(() => {
      dispatch(editPost(id, update));
    });
  };
};

const getAllPosts = posts => ({
  type: 'GET_ALL_POSTS',
  posts
});

const initiateGetAllPosts = () => {
  return (dispatch, getState) => {
    const { uid } = getState().auth;

    return database.ref(`users/${uid}/posts`).once('value').then((snapshot) => {
      const posts = [];

      snapshot.forEach((snapshotChild) => {
        posts.push({
          id: snapshotChild.key,
          ...snapshotChild.val()
        });
      });

      dispatch(getAllPosts(posts));
    });
  };
};

export {
  addPost,
  initiateAddPost,
  removePost,
  initiateRemovePost,
  editPost,
  initiateEditPost,
  getAllPosts,
  initiateGetAllPosts
};
