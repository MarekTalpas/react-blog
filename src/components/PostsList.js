import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import PostsListItem from './PostsListItem';
import filterPosts from '../selectors/posts';

export const PostsList = ({ posts }) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Posts</div>
      <div className="show-for-desktop">Posts</div>
    </div>
    <div className="list-body">
      {
        posts.length === 0 ? (
          <div className="list-item list-item--message">
            <span>No Posts To Show</span>
          </div>
        ) : (
          posts.map(post => <PostsListItem key={post.id} {...post} />)
        )
      }
    </div>
  </div>
);

PostsList.propTypes = {
  posts: PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    content: PropTypes.string.isRequired,
    createdAt: PropTypes.number.isRequired
  }).isRequired
};

const mapStateToProps = state => ({
  posts: filterPosts(state.posts, state.filters)
});

export default connect(mapStateToProps)(PostsList);
