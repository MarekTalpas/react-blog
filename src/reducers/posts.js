const defaultPostsState = [];

export default (state = defaultPostsState, action) => {
  switch (action.type) {
    case 'ADD_POST':
      return [
        ...state,
        action.post
      ];
    case 'REMOVE_POST':
      return state.filter(post => post.id !== action.id);
    case 'EDIT_POST':
      return state.map((post) => {
        if (post.id === action.id) {
          return {
            ...post,
            ...action.update
          };
        }
        return post;
      });
    case 'GET_ALL_POSTS':
      return action.posts;
    default:
      return state;
  }
};
