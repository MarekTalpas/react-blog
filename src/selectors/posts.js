import moment from 'moment';

export default (posts, {
  text,
  sortBy,
  startDate,
  endDate
}) => {
  return posts.filter((post) => {
    const creationDate = moment(post.createdAt);
    const textMatch = post.title.toLowerCase().includes(text.toLowerCase());
    const startDateMatch = startDate ? startDate.isSameOrBefore(creationDate) : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(creationDate) : true;

    return textMatch && startDateMatch && endDateMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return (a.createdAt < b.createdAt) ? 1 : -1;
    } else if (sortBy === 'title') {
      if (a.title.toLowerCase() < b.title.toLowerCase()) return 1;
      if (a.title.toLowerCase() > b.title.toLowerCase()) return -1;
      return 0;
    }
  });
};
