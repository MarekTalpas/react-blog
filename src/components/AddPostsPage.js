import React, { Component } from 'react';
import { connect } from 'react-redux';
import PostsForm from './PostsForm';
import { initiateAddPost } from '../actions/posts';

export class AddPostsPage extends Component {
  _handleFormSubmit = (post) => {
    this.props.initiateAddPost(post);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Create Post</h1>
          </div>
        </div>
        <div className="content-container">
          <PostsForm onSubmit={this._handleFormSubmit} />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  initiateAddPost: (post) => dispatch(initiateAddPost(post))
});

export default connect(undefined, mapDispatchToProps)(AddPostsPage);
