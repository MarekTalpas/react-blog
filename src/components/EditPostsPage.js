import React, { Component } from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import PostsForm from './PostsForm';
import { initiateEditPost, initiateRemovePost } from '../actions/posts';

window.Modal = Modal;

const modalStyles = {
  overlay: {
    display: 'flex',
    alignItems: 'center',
    height: '100vh',
    justifyContent: 'center',
    width: '100vw'
  },
  content: {
    borderRadius: '3px',
    boxShadow: '0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22)',
    margin: '$l-size $m-size',
    position: 'none',
    textAlign: 'center',
    width: '40rem'
  }
};

export class EditPostsPage extends Component {
  state = {
    modalIsOpen: false
  };
  _openModal = () => {
    this.setState(() => ({ modalIsOpen: true }));
  };
  _closeModal = () => {
    this.setState(() => ({ modalIsOpen: false }));
  };
  _handleFormSubmit = (post) => {
    this.props.initiateEditPost(this.props.post.id, post);
    this.props.history.push('/');
  };
  _handlePostRemove = () => {
    this.props.initiateRemovePost({ id: this.props.post.id });
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Post</h1>
          </div>
        </div>
        <div className="content-container">
          <PostsForm onSubmit={this._handleFormSubmit} post={this.props.post} />
          <button
            className="button button--secondary button--remove"
            onClick={this._openModal}
          >
            Remove Post
          </button>
        </div>
        <Modal
          isOpen={this.state.modalIsOpen}
          onRequestClose={this._closeModal}
          closeTimeoutMS={400}
          style={modalStyles}
          contentLabel="Remove Post"
        >
          <h1 className="box-layout__title">Do you really want to remove this post ?</h1>
          <div className="button__couple">
            <button
              className="button button--secondary button--confirm"
              onClick={this._closeModal}
            >
              NO
            </button>
            <button
              className="button button--confirm"
              onClick={this._handlePostRemove}
            >
              YES
            </button>
          </div>
        </Modal>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  post: state.posts.find(post => post.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
  initiateEditPost: (post) => dispatch(initiateEditPost(post)),
  initiateRemovePost: (id) => dispatch(initiateRemovePost(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPostsPage);
