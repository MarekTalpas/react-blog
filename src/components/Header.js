import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = ({ startLogout }) => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <Link className="header__title" to="/dashboard">
          <h1>React Blog</h1>
        </Link>
        <button className="button button--link" onClick={startLogout}>Logout</button>
      </div>
    </div>
  </header>
);

const mapDispatchToProps = dispatch => ({
  startLogout: () => dispatch(startLogout())
});

Header.propTypes = {
  startLogout: PropTypes.func.isRequired
};

export default connect(undefined, mapDispatchToProps)(Header);
