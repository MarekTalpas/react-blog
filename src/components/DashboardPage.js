import React from 'react';
import PostsListFilters from './PostsListFilters';
import PostsList from './PostsList';

const DashboardPage = () => (
  <div className="header--dashboard">
    <PostsListFilters />
    <PostsList />
  </div>
);

export default DashboardPage;
