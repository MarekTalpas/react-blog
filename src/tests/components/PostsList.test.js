import React from 'react';
import { shallow } from 'enzyme';
import { PostsList } from '../../components/PostsList';
import posts from '../fixtures/posts';

test('should render PostsList component with some posts properly', () => {
  const wrapper = shallow(<PostsList posts={posts} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render PostsList component with any posts properly', () => {
  const wrapper = shallow(<PostsList posts={[]} />);
  expect(wrapper).toMatchSnapshot();
});
