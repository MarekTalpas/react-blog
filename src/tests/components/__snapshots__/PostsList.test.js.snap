// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PostsList component with any posts properly 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Posts
    </div>
    <div
      className="show-for-desktop"
    >
      Posts
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No Posts To Show
      </span>
    </div>
  </div>
</div>
`;

exports[`should render PostsList component with some posts properly 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Posts
    </div>
    <div
      className="show-for-desktop"
    >
      Posts
    </div>
  </div>
  <div
    className="list-body"
  >
    <PostsListItem
      content="this is post about react and redux."
      createdAt={0}
      id="1"
      key="1"
      title="react and redux"
    />
    <PostsListItem
      content="this is post about javascript basics"
      createdAt={259200000}
      id="2"
      key="2"
      title="js basics"
    />
    <PostsListItem
      content="this is post about node js and all its features"
      createdAt={-259200000}
      id="3"
      key="3"
      title="nodejs"
    />
  </div>
</div>
`;
